contracts/FlashRollover.sol

```diff
@@ -0,0 +1,344 @@
+pragma solidity ^0.8.0;
+pragma experimental ABIEncoderV2;
+
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/token/ERC721/IERC721.sol";
+
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import "@openzeppelin/contracts/utils/math/SafeMath.sol";
+
+import "./external/interfaces/ILendingPool.sol";
+
+import "./interfaces/IFlashRollover.sol";
+import "./interfaces/ILoanCore.sol";
+import "./interfaces/IOriginationController.sol";
+import "./interfaces/IRepaymentController.sol";
+import "./interfaces/IAssetWrapper.sol";
+import "./interfaces/IFeeController.sol";
+
+contract FlashRollover is IFlashRollover {
+    using SafeERC20 for IERC20;
+    using SafeMath for uint256;
+
+    /**
+     * Holds parameters passed through flash loan
+     * control flow that dictate terms of the new loan.
+     * Contains a signature by lender for same terms.
+     * isLegacy determines which loanCore to look for the
+     * old loan in.
+     */
+    struct OperationData {
+        bool isLegacy;
+        uint256 loanId;
+        LoanLibrary.LoanTerms newLoanTerms;
+        uint8 v;
+        bytes32 r;
+        bytes32 s;
+    }
+
+    /**
+     * Defines the contracts that should be used for a
+     * flash loan operation. May change based on if the
+     * old loan is on the current loanCore or legacy (in
+     * which case it requires migration).
+     */
+    struct OperationContracts {
+        ILoanCore loanCore;
+        IERC721 borrowerNote;
+        IERC721 lenderNote;
+        IFeeController feeController;
+        IERC721 assetWrapper;
+        IRepaymentController repaymentController;
+        IOriginationController originationController;
+        ILoanCore newLoanLoanCore;
+        IERC721 newLoanBorrowerNote;
+    }
+
+    /* solhint-disable var-name-mixedcase */
+    // AAVE Contracts
+    // Variable names are in upper case to fulfill IFlashLoanReceiver interface
+    ILendingPoolAddressesProvider public immutable override ADDRESSES_PROVIDER;
+    ILendingPool public immutable override LENDING_POOL;
+    /* solhint-enable var-name-mixedcase */
+
+    // Pawn.fi Contracts
+    ILoanCore public immutable loanCore;
+    ILoanCore public immutable legacyLoanCore;
+    IOriginationController public immutable originationController;
```

Comment:
Not sure, but it might be more gas-effective to store these all in struct form? I believe that might save you a sload or 8

Reply:
Or do the suggestion above - store just the loanCore instances and fetch everything else from there